src/backend/libpq/rangerrest.c

```diff
@@ -0,0 +1,551 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*-------------------------------------------------------------------------
+ *
+ * rangerrest.c
+ *	routines to interact with Ranger REST API
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include <json-c/json.h>
+
+#include "utils/acl.h"
+#include "utils/guc.h"
+#include "utils/rangerrest.h"
+
+/*
+ * Internal buffer for libcurl context
+ */
+typedef struct curl_context_t
+{
+    CURL* curl_handle;
+
+    char curl_error_buffer[CURL_ERROR_SIZE];
+
+    int curl_still_running;
+
+    struct
+    {
+        char* buffer;
+        int size;
+    } response;
+
+    char* last_http_reponse;
+} curl_context_t;
+typedef curl_context_t* CURL_HANDLE;
+
+RangerACLResult parse_ranger_response(char* buffer)
+{
+    Assert(buffer != NULL);
+    if (strlen(buffer) == 0)
+        return RANGERCHECK_UNKNOWN;
+
+    elog(LOG, "read from Ranger Restful API: %s", buffer);
+
+    struct json_object *response = json_tokener_parse(buffer);
+    struct json_object *accessObj = json_object_object_get(response, "access");
+
+    int arraylen = json_object_array_length(accessObj);
+    elog(LOG, "Array Length: %dn",arraylen);
+
+    json_object * jvalue;
+    json_object * jallow;
+    json_bool result;
+    // here should return which table's acl check failed in future.
+    for (int i=0; i< arraylen; i++){
+      jvalue = json_object_array_get_idx(accessObj, i);
+      jallow = json_object_object_get(jvalue, "allowed");
+      result = json_object_get_boolean(jallow);
+      if(result != 1){
+        return RANGERCHECK_NO_PRIV;
+      }
+    }
+    return RANGERCHECK_OK;
+
+}
+
+/*
+ * A mapping from AclObjectKind to string
+ */
+char* AclObjectKindStr[] =
+{
+    "table",             /* pg_class */
+    "sequence",          /* pg_sequence */
+    "database",          /* pg_database */
+    "function",          /* pg_proc */
+    "operator",          /* pg_operator */
+    "type",              /* pg_type */
+    "language",          /* pg_language */
+    "namespace",         /* pg_namespace */
+    "oplass",            /* pg_opclass */
+    "conversion",        /* pg_conversion */
+    "tablespace",        /* pg_tablespace */
+    "filespace",         /* pg_filespace */
+    "filesystem",        /* pg_filesystem */
+    "fdw",               /* pg_foreign_data_wrapper */
+    "foreign_server",    /* pg_foreign_server */
+    "protocol",          /* pg_extprotocol */
+    "none"               /* MUST BE LAST */
+};
```

Comment:
a few of these are not supported in RPS/Ranger: operator, type, namespace, oplass, conversion, filespace, filesystem, fdw, foreign_server. Will these types be submitted to RPS or will they be mapped to other types ?